### DISCUSSION TITLE
update_credentials API deletes fields when passed field_attributes []

### DISCUSSION THREAD

**Discussion by dickhardt** [mastodon/mastodon#22146] [Q&A] [answered]
per https://docs.joinmastodon.org/methods/accounts/#update_credentials

I'm passing a JSON object with `fields_attributes`  (which is weird, since that is not the name of the attribute in the account object

The command
```
curl -v -X PATCH \
    -d '{"fields_attributes": [ { "name": "Foo", "value": "Bar" } ]}' \
    -H 'Content-Type: application/json' \
    -H 'Authorization: Bearer <ACCESS_TOKEN>' \
    https://mastodon.example/api/v1/accounts/update_credentials
```
does not update any fields, and deletes any existing fields

This command is working fine
```
curl -v -X PATCH \
    -d '{"display_name": "Test User"' \
    -H 'Content-Type: application/json' \
    -H 'Authorization: Bearer <ACCESS_TOKEN>' \
    https://mastodon.example/api/v1/accounts/update_credentials
```

**Comment by trwnh** [comment, marked as answer]
It's `field_attributes` in the account model:

https://github.com/mastodon/mastodon/blob/ed07f10ca8d4e65ec58958f300a8bb7c762ccbbd/app/models/account.rb#L305-L325

In any case, it turns out you need to include the index for the fields_attributes. In other words, use `fields_attributes[0][name]` instead of `fields_attributes[][name]`. For example:

- Not working: `http://mastodon.local/api/v1/accounts/update_credentials?fields_attributes[][name]=Foo&fields_attributes[][value]=Bar`
- Working: `http://mastodon.local/api/v1/accounts/update_credentials?fields_attributes[0][name]=Foo&fields_attributes[0][value]=Bar`

I'm not sure why this API method is different from other API methods, though...

It seems that for the JSON version of the API, you have to pass it in as such:

```json
{
  "fields_attributes": {
    "0": {
      "name": "Foo",
      "value": "Bar"
    }
  }
}
```

I don't know why it's like this.

**Reply by dickhardt** [reply]
Thanks! That works!
